lib/Doctrine/DBAL/Internal/CommitOrderCalculator.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+namespace Doctrine\DBAL\Internal;
+
+use function array_reverse;
+
+/**
+ * CommitOrderCalculator implements topological sorting, which is an ordering
+ * algorithm for directed graphs (DG) and/or directed acyclic graphs (DAG) by
+ * using a depth-first searching (DFS) to traverse the graph built in memory.
+ * This algorithm have a linear running time based on nodes (V) and dependency
+ * between the nodes (E), resulting in a computational complexity of O(V + E).
+ */
+class CommitOrderCalculator
```

Comment:
Although this is used for commit order calculation, it's not specialized for that. Would it make sense to rename this class and the related ones in a way that doesn't imply that it's only for commits? E.g. `Graph`, `Node`, `Edge`?

Reply:
We should make it `final` too

Reply:
Perhaps we could bring this to `doctrine/persistence` to be reuse by the orm too?

Reply:
I made it final now, we can move it to persistence in another step, but ORM depends on DBAL, so it can already use it?

Reply:
If it's needed in ORM together with the methods like `hasNode()` that are not used by DBAL and probably weights as well, why not move it to a shared library instead?

It feels weird that the code is ported from one library to another as is just because it's needed as is for the library it's currently located in.

Reply:
@morozov i don't see how its weird, DBAL is the common denominator of where this code is needed in terms of dependencies. Introducing yet another library is just adding overhead. In addition it would prevent this fix from being merged into 2.10 per the dependencies are locked rule.

Reply:
IMO, the "common denominator" approach is wrong and counterproductive. This way, the consumers of the shared logic end up storing and maintaining the logic they need in a place where it doesn't belong.

Eventually, the changes in this logic are driven by the consumers while the owners of the library end up accepting everything "just because it's needed for project X".

In this case, I see two proper solutions:

1. Have `@internal` implementations that solve the problems of each of the libraries independently and evolve independently (looks the most reasonable in this case).
2. Have a shared library that solves this problem in general (look like overkill in this case).

Reply:
We can keep this code as a 100% copy of the ORM code. This is not an attempt to force the ORM to use the class from here in the future. I just happened to need the exact same logic, because its the same problem, hence suggesting it can be re-used by the ORM in the future.

Reply:
I disagree. The commit order resolution is an _internal_ (as the namespace suggests) implementation detail of the SQLite driver which is _not_ part of the database abstraction layer. It cannot be planned for reuse by another component. And this argument cannot be used to keep the pieces of the code that this component doesn't need for its _internal_ purposes.